redact/index.ts

```diff
@@ -0,0 +1,210 @@
+};
+
+function userEntitiesToWasm(entity: unknown) {
+  if (typeof entity !== "string") {
+    throw new Error("Redaction entities must be a string");
+  }
+
+  if (entity === "email") {
+    return { tag: "email" as const };
+  }
+
+  if (entity === "phone-number") {
+    return { tag: "phone-number" as const };
+  }
+
+  if (entity === "ip-address") {
+    return { tag: "ip-address" as const };
+  }
+
+  if (entity === "credit-card-number") {
+    return { tag: "credit-card-number" as const };
+  }
+
+  return {
+    tag: "custom" as const,
+    val: entity,
+  };
+}
+
+function wasmEntitiesToString(entity: SensitiveInfoEntity) {
+  if (entity.tag === "email") {
+    return "email";
+  }
+
+  if (entity.tag === "ip-address") {
+    return "ip-address";
+  }
+
+  if (entity.tag === "credit-card-number") {
+    return "credit-card-number";
+  }
+
+  if (entity.tag === "phone-number") {
+    return "phone-number";
+  }
+
+  return entity.val;
+}
+
+function performReplacementInText(
+  text: string,
+  replacement: string,
+  start: number,
+  end: number,
+): string {
+  return text.substring(0, start) + replacement + text.substring(end);
+}
+
+function noOpDetect(_tokens: string[]): Array<SensitiveInfoEntity | undefined> {
+  return [];
+}
+function noOpReplace(_input: SensitiveInfoEntity): string | undefined {
+  return undefined;
+}
+
+interface RedactedSensitiveInfoEntity {
+  original: string;
+  redacted: string;
+  start: number;
+  end: number;
+  identifiedType: string;
+}
+
+async function callRedactWasm<
+  const Detect extends DetectSensitiveInfoEntities<CustomEntities> | undefined,
+  const CustomEntities extends string,
+>(
+  candidate: string,
+  options?: RedactOptions<Detect>,
+): Promise<RedactedSensitiveInfoEntity[]> {
+  let convertedDetect = noOpDetect;
+  if (typeof options?.detect === "function") {
+    const detect = options.detect;
+    convertedDetect = (tokens: string[]) => {
+      return detect(tokens)
+        .filter((e) => typeof e !== "undefined")
+        .map((e) => userEntitiesToWasm(e));
+    };
+  }
+
+  let convertedReplace = noOpReplace;
+  if (typeof options?.replace === "function") {
+    const replace = options.replace;
+    convertedReplace = (identifiedType: SensitiveInfoEntity) => {
+      return replace(
+        // @ts-ignore because we know this is coming from Wasm
+        wasmEntitiesToString(identifiedType),
+      );
+    };
+  }
+
+  const wasm = await initializeWasm(convertedDetect, convertedReplace);
+
+  if (typeof wasm !== "undefined") {
+    const skipCustomDetect = typeof options?.detect !== "function";
+    const skipCustomRedact = typeof options?.replace !== "function";
+
+    if (options?.entities?.length === 0) {
+      console.warn("no entities configured for redaction");
+    }
+
+    const config = {
+      entities: Array.isArray(options?.entities)
+        ? options.entities.map(userEntitiesToWasm)
+        : undefined,
```

Comment:
Now that you are checking the length, we can also do the `Array.isArray` check above (before you check the length). You need to check it's an array before you check the length because objects, strings, etc can all have length properties.

Reply:
I've tidied this up a little bit and made it throw